util/iostat/iostat.go

```diff
@@ -0,0 +1,132 @@
+package iostat
+
+import (
+	"bufio"
+	"context"
+	"errors"
+	"fmt"
+	"os/exec"
+	"strconv"
+	"strings"
+
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/metrics"
+)
+
+type MetricsSpawner struct {
+	statReceiver chan DeviceStats
+	metrics      map[string]map[string]metrics.GaugeFloat64
+}
+
+func NewMetricsSpawner() *MetricsSpawner {
+	return &MetricsSpawner{
+		metrics:      make(map[string]map[string]metrics.GaugeFloat64),
+		statReceiver: make(chan DeviceStats),
+	}
+}
+
+func (m *MetricsSpawner) RegisterMetrics(ctx context.Context, spwanInterval int) error {
+	go Run(ctx, spwanInterval, m.statReceiver)
+	// Register metrics for a maximum of 5 devices (fail safe incase iostat command returns incorrect names indefinitely)
+	for i := 0; i < 5; i++ {
```

Comment:
After reading the code, I imagine that if the machine has a single device, it will have to wait at least 5 seconds to receive the stat from the same device before being able to start. 

Instead of limiting the number of devices, you could register them on demand on the PopulateMetrics function with the function [GetOrRegisterGaugeFloat64](https://github.com/ethereum/go-ethereum/blob/v1.14.5/metrics/gauge_float64.go#L20).

Reply:
`RegisterMetrics` generally should only run for 1 second (the time interval supplied to the iostat command where it reports all devices info all at once every second) and as soon as a repetition in device name is seen we conclude that registration was successful.

Running the loop for 5 iterations is to limit metrics registration to at most 5 devices (this can be made configurable), so that in case we get erroneous data from the command's output we don't keep registering new metrics indefinitely, GetOrRegisterGaugeFloat64 wouldn't solve this problem.

Reply:
> RegisterMetrics generally should only run for 1 second (the time interval supplied to the iostat command where it reports all devices info all at once every second) and as soon as a repetition in device name is seen we conclude that registration was successful.

Yeah, I missed that. However, I'm still slightly bothered by this extra one-second delay during initialization. Couldn't you move this initialization logic to PopulateMetrics?

Also, it would be good to have this value configurable.

Reply:
merged the two functions and if there are more devices than maxDeviceCount, they are simply ignored. Will make this number configurable based on the response from sre.